transformer_engine/common/include/transformer_engine/comm_gemm_overlap.h

```diff
@@ -0,0 +1,201 @@
+/*************************************************************************
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See LICENSE for license information.
+ ************************************************************************/
+
+#ifndef TRANSFORMER_ENGINE_COMMON_COMM_GEMM_OVERLAP_H_
+#define TRANSFORMER_ENGINE_COMMON_COMM_GEMM_OVERLAP_H_
+
+#include <cuda.h>
+#include <cuda_fp8.h>
+#include <transformer_engine/transformer_engine.h>
+
+#include <functional>
+
+#include "common/comm_gemm_overlap/userbuffers/userbuffers.h"
+
+#define NVTE_COMM_OVERLAP_MAX_STREAMS 3
+
+namespace transformer_engine {
+
+bool device_supports_multicast();
+
+bool ubuf_built_with_mpi();
+
+enum class CommOverlapType { RS = 0, AG = 1 };
+
+enum class CommOverlapAlgo {
+  BULK_OVERLAP_AG = 0,
+  BULK_OVERLAP_RS = 1,
+  SPLIT_PIPELINED_AG_P2P = 2,
+  SPLIT_PIPELINED_RS = 3,
+  SPLIT_PIPELINED_RS_P2P = 4,
+  ATOMIC_GEMM_RS = 5,
+  ATOMIC_GEMM_AG_P2P = 6,
+  ATOMIC_GEMM_RS_P2P = 7
+};
+
+class CommOverlapCore {
```

Comment:
[`TensorWrapper` explicitly deletes copy-constructors](https://github.com/NVIDIA/TransformerEngine/blob/d9b4bfb5a20b1b0b3bc69c50e33b78ccfdef9e72/transformer_engine/common/include/transformer_engine/transformer_engine.h#L225-L226) so we need to use rvalue semantics in the function signatures, but I agree that the arguments we modify need to be `TensorWrapper&&` instead of `const TensorWrapper&&`. I just forgot that `&&` existed. `TensorWrapper*` is also a possibility but it makes the call statements a big ugly on the framework side.

As for the member variables, I left everything public in the interest of not restricting what the framework wrappers can do with these classes, but then it makes more sense for them to be structs instead of classes anyway. I'll review what we can protect and/or make private, and then expose selectively for frameworks as the need arises.

Reply:
Passing an rvalue into an output arg seems weird to me since there's no guarantee they're valid when the function exits. It feels like `TensorWrapper&` is the right type. I personally don't agree with cpplint about it, so it might be worth just suppressing that warning (or maybe we can remove that from cpplint's rules).

Reply:
I think making all the args `TensorWrapper&` is much better. In the future, after we take more consideration of const correctness, we can make the input args `const TensorWrapper&`.